pkg/applicationserver/io/packages/alcsync/v1/package.go

```diff
@@ -0,0 +1,111 @@
+// Copyright © 2023 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package alcsyncv1
+
+import (
+	"context"
+
+	"go.thethings.network/lorawan-stack/v3/pkg/applicationserver/io"
+	"go.thethings.network/lorawan-stack/v3/pkg/applicationserver/io/packages"
+	"go.thethings.network/lorawan-stack/v3/pkg/errors"
+	"go.thethings.network/lorawan-stack/v3/pkg/log"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+)
+
+// PackageName is the name of the package.
+const PackageName = "alcsync-v1"
+
+var errNoAssociation = errors.DefineInternal("no_association", "no association available")
+
+type alcsyncpkg struct {
+	server   io.Server
+	registry packages.Registry
+}
+
+// HandleUp implements packages.ApplicationPackageHandler.
+func (a *alcsyncpkg) HandleUp(
+	ctx context.Context,
+	def *ttnpb.ApplicationPackageDefaultAssociation,
+	assoc *ttnpb.ApplicationPackageAssociation,
+	up *ttnpb.ApplicationUp,
+) error {
+	ctx = log.NewContextWithField(ctx, "namespace", "applicationserver/io/packages/alcsync/v1")
+	logger := log.FromContext(ctx)
+
+	if def == nil && assoc == nil {
+		logger.Error("No association available")
+		return errNoAssociation.New()
+	}
+
+	logger.Debug("Handle uplink")
+
+	msg := up.GetUplinkMessage()
+	if msg == nil {
+		logger.Debug("Uplink is not an uplink message")
+		return nil
+	}
+
+	data, err := mergePackageData(def, assoc)
+	if err != nil {
+		logger.Error("Failed to merge package data")
+		return err
+	}
+
+	if msg.GetFPort() != data.FPort {
+		logger.Debug("Uplink received on unhadled FPort")
+		return nil
+	}
+
+	if len(msg.GetFrmPayload()) == 0 {
+		logger.Debug("Uplink message has no payload")
+		return nil
+	}
+
+	commands, err := MakeCommands(msg, data)
+	if err != nil {
+		logger.WithError(err).Debug("Failed to parse frame payload into commands")
+		return err
+	}
+	for _, cmd := range commands {
+		downlinks, err := cmd.Execute()
+		if err != nil {
+			logger.WithError(err).Debug("Failed to execute command")
+			continue
+		}
+		if len(downlinks) == 0 {
+			continue
+		}
+		if err := a.server.DownlinkQueuePush(ctx, up.EndDeviceIds, downlinks); err != nil {
```

Comment:
You may benefit from accumulating them from each command and then doing a singular push after the loop. Less round trips, as `Push` needs to call the NS.